### ISSUE TITLE
$TMP $TEMP do not contain the content of %TMP% nor %TEMP%

### ISSUE THREAD

**Issue by dumblob** [git-for-windows/git#1687]
First of all let me express my respect to this project. 3 years ago there was no portable (nearly all Windows versions, small size, guaranteed maintainability, known standardized tools and languages and environment in general) sane way of scripting on Windows (PowerShell was more or less close, but not enough). This project allowed exactly that even though it's primarily focused on git.

Thank you very much for the hard work and to all the guys from MinGW and MSYS2 as they provided the basis for this to happen.

 - [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

I found just related issues: https://github.com/git-for-windows/git/issues/89 , https://github.com/git-for-windows/git/issues/123 , https://github.com/git-for-windows/git/issues/143

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options
git version 2.13.1.windows.1
built from commit: 13f583bedb59f9a11b3f2e91673c10c250f84576
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver
Microsoft Windows [Version 10.0.15063]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt
Path Option: Cmd
SSH Option: OpenSSH
CURL Option: WinSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

```
$ echo $TMP $TEMP
/tmp /tmp
$ cmd.exe /c "echo %TMP% %TEMP%"
C:\Users\user00\AppData\Local\Temp C:\Users\user00\AppData\Local\Temp
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

`bash`

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

Different Windows applications (e.g. `rc.exe`, `cl.exe`, `link.exe`).

 - What did you expect to occur after running these commands?

It's not specific to any run of a command.

I expected `$TMP` and `$TEMP` to be set exactly to the same values as `%TMP%` and `%TEMP%`.

 - What actually happened instead?

The `$TMP` and `$TEMP` variables are not set to the same values (i.e. including backslashes instead of forward slashes) as `%TMP%` and `%TEMP%`, which makes them useless for Windows scripting (**Windows apps have no clue what a path beginning with a forward slash nor backslash means**).

Now I need to use a workaround - `TMP="$(cmd.exe /c "echo %TMP%")"` and `TEMP="$(cmd.exe /c "echo %TEMP%")"` which feels kind of awkward 😢 .

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

N/A

Could we just fill `$TMP` and `$TEMP` with the values of `%TMP%` and `%TEMP%` respectively (i.e. they might stay probably with backslashes)? Of course the `/tmp` directory will stay untouched (i.e. a virtual mount to `%TMP%`).

**Comment by leonyu** [comment]
I believe this is a MSYS2/Cygwin problem. 

There was [this thread](http://cygwin.1069669.n5.nabble.com/TMP-and-TEMP-get-redefined-but-I-need-the-original-values-tt31231.html) in Cygwin mailing list in 2011 discussing this change.

TL;DR Windows file system doesn't correctly persist Unix permission, thus breaking some programs.

**Comment by dumblob** [comment]
@leonyu thank you for the pointer. That's quite interesting.

I would anyway argue, that it's about 8 years since `%TMP%` and `%TEMP%` are not being used in Cygwin any more and therefore the need for it doesn't sound nowadays that valid to me. I would guess the issue with permissions occured on older platforms (e.g. Win XP) and I would definitely give it a try on newer platforms (Win 7 & Win 10). There might have been also some change in `patch` since then as `patch` seems to have been the motivating example for not using `%TMP%` and `%TEMP%`.

I could not find any reference to a more or less recent Cygwin tests in this regard on Win 7 and Win 10. Have you found any, @leonyu ?

**Comment by leonyu** [comment]
Actually, I just checked, in recent version of Git, `/tmp` is a mount of %TEMP%. It just that it is done via Cygwin mount point, so that it is transparent to most UNIX programs, and masks permission correctly. You can get the Windows path by running `cygpath -m /tmp`

**Comment by dumblob** [comment]
@leonyu thanks for the tip with `cygpath` (does it return a unix path with forward slashes or really the very original windows path with backward slahes?). The whole point of this issue has though nothing to do with virtual mount points, but rather with using `$TMP` and `$TEMP` with native Windows applications as I pointed out in the issue description.

It's also really confusing mounting `%TEMP%` to `/tmp` as I would definitely expect `%TEMP%` to be mounted to `/temp` and rather `%TMP%` to be mounted to `/tmp`.

I would really like to see any current real examples of issues with the permissions, which would defend the current Cygwin defaults. I could not find any. If there aren't any, I might consider trying it myself - e.g. the behavior of `patch` on Windows 10.

**Comment by leonyu** [comment]
I think you can get backward slash version by using `cygpath -w /tmp`. 

`%TMP%` and `%TEMP%` are both Windows environment variable. They are not Unix env variables, the UNIX variable for tmp directory should be `$TMPDIR` if it is defined, though a lot of places don't define it.

https://en.wikipedia.org/wiki/TMPDIR

As for trying to to unmap tmp, good luck, lol.

**Comment by dumblob** [comment]
@leonyu thanks for the tip with `cygpath -w /tmp`, it might be helpful in some situations. Still it doesn't seem to be related to this issue.

>`%TMP%` and `%TEMP%` are both Windows environment variable. They are not Unix env variables, the UNIX variable for tmp directory should be `$TMPDIR` if it is defined, though a lot of places don't define it.

Yes, of course. But how does information relate to this issue? I've explicitly stated that I'm interested in running native Windows applications (`rc.exe` `cl.exe` `link.exe` ...) which do need one of the standard Windows temporary directories which are accessible through `%TMP%` and `%TEMP%`. In git-for-windows or Cygwin respectively there are equivalents exactly for this purpose (`$TMP` and `$TEMP`), but they both point to different directories :open_mouth: than their Windows counterparts and even are not in any way compatible with native Windows applications (just try to use the string `/tmp` or `\tmp` as a directory argument with these Windows native applications and you'll see what I mean).

In other words a sole Windows-specific stuff available through unix environment "wrapper" with the sole purpose to not have to run hacks like `TMP="$(cmd.exe /c "echo %TMP%")"` is so screwed, that it's even worse than not having `$TMP` and `$TEMP` at all (because not having it would be easily detectable, but right now it has severe consequences as the Windows native apps using Windows path routines will understand `/tmp` as `\tmp` which means `C:\tmp` which is really not nice - permissions and security, available space, etc. - all that kind of stuff is gone, not speaking about company policies on personal computers as well as servers...).

>As for trying to to unmap tmp, good luck, lol.

Not sure what you mean. Where did you get the idea to unmap `/tmp`? Why would anyone want to do that to solve the issue? I'm asking especially because I mentioned explicitly in the issue description at the bottom, that `/tmp` shall stay untouched.

**Comment by dscho** [comment]
First of all, please note that this here bug tracker is actually meant to report bugs *in Git for Windows*.

In other words, if you use Git for Windows for other purposes than running Git, you may get lucky, but when things start not working, you're really on your own. The ratio between users of Git for Windows *users* and *developers* (or for that matter even just interested users who help other users) is really *so* high that we cannot take on user questions beyond Git.

Having said that, it is my understanding that calling pure Win32 applications from Git Bash will automagically convert paths and environment variables from Unix-y paths to Windows-y paths. Example: `git config -e -f /tmp/a1` will open the correct file. Even if the editor is a Win32 program such as VS Code.

As to `%TMP%` referring to something different than `$TMP`: I canno reproduce that behavior. In fact, as long as users do not fiddle with internals, `/tmp` is mounted to the *very exact* location that `%TMP%` refers to.

So I am convinced that the reported issue is not really a real bug.

**Comment by dumblob** [comment]
@dscho thanks for the reply.

I'm having a feeling we're talking slightly at cross purposes. My issue is, that `$TMP` is not set to the exact string contained in `%TMP%` (same goes for `$TEMP` as it contains a different string than `%TEMP%`).

I'm also quite certain that this leads to issues with invocation of shell in conjunction with git (i.e. this repository is the correct place to report this issue) as one does need to invoke Win32 apps directly (i.e. not through git itself like your example) as I showed in the issue description (`cl.exe` is a Visual Studio compiler, `link.exe` is a Visual Studio linker, etc.).

I can post a trivial patch for this issue, but first I would prefer reaching a consensus in the git-for-windows community that it's the right thing to do.

The current state leads to huge subtle bugs which are quite difficult to debug and have some severe effects. It's really hard to catch it as using native shell apps and capabilities to debug it shows everything is correct as they all can work with forward slashes in a string and thanks to the auto conversion from backward slashes they work indistinguishably correct also with paths with backward slashes).

**Comment by dscho** [comment]
> My issue is, that `$TMP` is not set to the exact string contained in `%TMP%` (same goes for `$TEMP` as it contains a different string than `%TEMP%`).

Neither is `$PATH` set to the exact string contained in `%PATH%`. Or for that matter, `$ProgramFiles`, `$SystemRoot` and probably many others.

The point is: yes, MSYS2 Bash converts environment variables that look as if they refer to paths or path lists. That is very much by design.

**Comment by dumblob** [comment]
>Neither is  `$PATH` set to the exact string contained in `%PATH%`. Or for that matter, `$ProgramFiles`, `$SystemRoot` and probably many others.
>
>The point is: yes, MSYS2 Bash converts environment variables that look as if they refer to paths or path lists. That is very much by design.

Yes, you're very much correct. The point is, that the semantics in MSYS2 design seems to be inconsistent when doing this conversion (there seems to be different exceptions, etc.). My understanding is, that generally the goal in MSYS2 is to convert these paths "at the last mile" (i.e. not in variables, be it an environment variable or not, but rather first at the very last moment before passing it as an argument to an executable).

More importantly though, strictly Windows-specific variables (e.g. `%TMP%` and `%TEMP%`, but not those with name clash like `%PATH%`) are expected to not get converted (or when converted, then also renamed to avoid name clashes).

**Comment by PhilipOakley** [comment]
Hi @dumblob,

> I can post a trivial patch for this issue, but first I would prefer reaching a consensus in the git-for-windows community that it's the right thing to do.

A PR/patch would provide a solid base to clarify an misunderstandings and makes it easier to accept incremental improvements.

**Comment by dscho** [comment]
> strictly Windows-specific variables (e.g. `%TMP%` and `%TEMP%`, but not those with name clash like `%PATH%`)

The `$TMP`/`$TEMP` variables can be used in Bash scripts, so...
